data_transfer/dags/btf.py

```diff
@@ -17,24 +25,59 @@ def dag(study_site: StudySite) -> None:
         ->batch_upload_data
 
+    Parameters
+    ----------
+    study_site : StudySite
+    days : int, optional
+        The amount of days in the past to query data for. This defaults to 50 days
+        to ensure prior data which has been delayed in upload is picked up
+    delta : int, optional
+        Initial reference (0 == today) to query backwards from. This allows
+        traversing back in time for historical data
+
+
     NOTE/TODO: this method simulates the pipeline.
     """
+    byteflies = Byteflies(study_site)
+
+    data_period = get_period_by_days(int(delta), int(days))
 
-    # TODO: ensure that we get the End of Day as intended based on automation schedule
-    # TODO: should we ensure overlap with previous run? How often do we run this DAG?
-    # TODO: perhaps pull this one abstraction higher (into the init of this DAG?)
-    data_period = get_period_by_days(datetime.today(), 2)  # NOTE: two days for testing
-    byteflies_jobs.batch_metadata(*data_period, study_site)
+    log.debug(
+        f"Dowloading records from {datetime.fromtimestamp(int(data_period[0]))}"
+        f" to {datetime.fromtimestamp(int(data_period[1]))}"
+    )
+
+    byteflies_jobs.batch_metadata(byteflies, *data_period)
 
     results = records_not_uploaded(DeviceType.BTF)
 
-    for _patient_device, records in results.items():
+    # NOTE: group records by patients per device to process small batches.
+    for patient_device, records in results.items():
         for record in records:
             # Each task should be idempotent. Returned values feeds subsequent task
-            mongoid = byteflies_tasks.task_download_data(record.id)
+            mongoid = byteflies_tasks.task_download_data(byteflies, record.id)
             byteflies_tasks.task_preprocess_data(mongoid)
 
-        # Data is finalised and moved to a folder in /uploading/
-        shared_jobs.prepare_data_folders(DeviceType.BTF)
-        # All said folders FOR ALL DEVICES are uploaded once per day
-        shared_jobs.batch_upload_data(DeviceType.BTF)
+        # Only upload when all records are ready
+        if all_records_downloaded(records):
+            log.debug(f"All records for {patient_device} DOWNLOADED -> PREPARING ...")
+            shared_jobs.prepare_data_folders(DeviceType.BTF)
+            log.debug(f"All records for {patient_device} PREPARED   -> UPLOADING ...")
+            shared_jobs.batch_upload_data(DeviceType.BTF)
+        else:
+            log.error(f"Some records for {patient_device} were not downloaded.")
+
+
+def historical_dag(
+    study_site: StudySite, _days: Union[int, str] = -1, delta: Union[int, str] = 0
+) -> None:
+    """Loops through set periods to retreive all historical data"""
+    days_to_cover = (
+        datetime.today() - datetime.fromisoformat(config.byteflies_historical_start)
+    ).days
+    # jump straight to delta reference and skip days in between
+    tracker = int(delta)
+    while tracker < days_to_cover:
```

Comment:
How come `tracker` is defined? You could use `delta` in the while loop in the same way?

Reply:
Good point!